dist_run.py

```diff
@@ -6,67 +6,151 @@
+from build.model import TransformerArgs
 from build.model_dist import TransformerStage
+from distributed.logging_utils import setup_logging
+from distributed.safetensor_utils import (
+    get_hf_config_file,
+    get_hf_weight_map_and_path,
+    load_safetensor_weights,
+)
+from distributed.dtensor_utils import find_cpu_tensors, record_module_dtypes
+from distributed.utils import Color as color
+
+MODEL_NAME = "Transformer-2-7b-chat-hf"
+NAME_TO_HF_MODEL_ID_AND_DTYPE = {
+    "Transformer-2-7b-chat-hf": ("meta-llama/Llama-2-7b-chat-hf", torch.float16),
+}
+
+
+def _init_distributed():
+    dist.init_process_group("nccl")
+    rank = dist.get_rank()
+    world_size = dist.get_world_size()
+    torch.cuda.set_device(rank)
+    return rank, world_size
+
+
+def _create_device_mesh(mesh_dimensions):
+    return dist.init_device_mesh("cuda", mesh_dimensions, mesh_dim_names=("pp", "tp"))
+
+
+def _load_model_weights(stage_module, hf_model_name, device, logger):
+    weight_map, weight_path, key_map = get_hf_weight_map_and_path(hf_model_name)
+    num_loaded_weights, num_missing_weights = load_safetensor_weights(
+        stage_module, weight_map, weight_path, key_map, device
+    )
+    logger.info(
+        f"Success - Loaded {num_loaded_weights} weights, {num_missing_weights} missing weights"
+    )
+    if num_missing_weights > 0:
+        raise ValueError(f"Missing {num_missing_weights} weights")
+
+
+def _cleanup():
+    dist.barrier()
+    dist.destroy_process_group()
+
 
-# Model config
 def main():
-    config = ModelArgs.from_name("Transformer-2-7b-chat-hf").text_transformer_args
-    print(config)
+    rank, world_size = _init_distributed()
+    logger = setup_logging(__name__)
+
+    config = TransformerArgs.from_name(MODEL_NAME)
+    logger.info(f"Chat Model Config: {config}")
+
+    hf_model_name, model_dtype = NAME_TO_HF_MODEL_ID_AND_DTYPE[MODEL_NAME]
+    logger.info(f"Using HF model weights from {hf_model_name} and dtype {model_dtype}")
+
+    hf_config = get_hf_config_file(hf_model_name)
+    if hf_config is None:
+        raise ValueError(f"Config file not found for model id {hf_model_name}")
+    logger.info(f"Using HF model weights from {hf_model_name}")
+
+    mesh_dimensions = (2, 2)
+    device_mesh = _create_device_mesh(mesh_dimensions)
 
-    # Construct a device mesh with available devices (multi-host or single host)
-    device_mesh = dist.init_device_mesh("cuda", (2, 2), mesh_dim_names=("pp", "tp"))
     tp_mesh = device_mesh["tp"]
     pp_mesh = device_mesh["pp"]
     pp_rank = pp_mesh.get_local_rank()
     nstages = pp_mesh.size()
-
-    rank = dist.get_rank()
     device = torch.device(f"cuda:{rank}")
 
-    # Create parallel model with device_mesh context
     with device:
         with tp_mesh:
             model = TransformerStage(config, pp_rank, nstages)
             model.setup_caches(1, 4096)
+            # TODO: refine this .to once we start using fp8 for KV cache
+            model = model.to(model_dtype)
```

Comment:
nit: 
1. see if `config` provides an option for dtype.
2. Note that `build.utils` has a function named `get_precision()`. It seems to me that's where all precisions are controlled.
![Screenshot 2024-08-29 at 3 39 43 PM](https://github.com/user-attachments/assets/631f5a2c-2b2a-4d95-90e4-3a2613367593)

Reply:
good ideas but atm neither do what we need.  
1 - config does not contain model dtype
2 - tested the get precision api, and just returns float32 atm, so I've left a commented out line using it with a TODO to investigate.  It would be nice if this works so we don't have to maintain our own list. 

~~~
[rank0]:08-29 19:51:19.455 - __main__:65 - INFO - Chat Model Config: ModelArgs(text_transformer_args=TransformerArgs(block_size=2048, vocab_size=32000, n_layers=32, n_heads=32, dim=4096, hidden_dim=11008, n_local_heads=32, head_dim=128, rope_base=10000, norm_eps=1e-05, multiple_of=256, ffn_dim_multiplier=None, use_tiktoken=False, max_seq_length=8192, use_scaled_rope=False))
[rank0]:08-29 19:51:19.455 - __main__:67 - INFO - Using torchchat precision torch.float32
~~~